include/revng/PipelineC/Tracing/TraceFile.h

```diff
@@ -0,0 +1,154 @@
+#pragma once
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <string>
+#include <vector>
+
+#include "llvm/Support/Error.h"
+#include "llvm/Support/YAMLTraits.h"
+
+#include "revng/Support/Assert.h"
+
+namespace revng::tracing {
+
+enum ArgumentState {
+  Undefined,
+  Scalar,
+  Sequence,
+};
+
+struct Argument {
+private:
+  ArgumentState State;
+
+public:
+  std::string Scalar;
+  std::vector<std::string> Sequence;
+
+  Argument() : State(Undefined), Scalar(), Sequence() {}
+
+  void SetState(const ArgumentState NewState) {
+    revng_assert(State == Undefined || State == NewState);
+    State = NewState;
+  }
+
+  bool verify() const { return State != ArgumentState::Undefined; }
+  bool isScalar() const { return State == ArgumentState::Scalar; }
+  bool isSequence() const { return State == ArgumentState::Sequence; }
+};
+
+struct Command {
+public:
+  std::string Name;
+  std::vector<Argument> Arguments;
+  std::string Result;
+};
+
+struct Trace {
+public:
+  uint64_t Version;
+  std::vector<Command> Commands;
+
+  static llvm::Expected<Trace> fromFile(const llvm::StringRef Path);
```

Comment:
[B.16](https://pad.rev.ng/s/SJNUet_ku#B-CC)
```suggestion
public:
  static llvm::Expected<Trace> fromFile(const llvm::StringRef Path);
```